src/python/pants/core/goals/generate_lockfiles.py

```diff
@@ -124,6 +140,156 @@ class RequestedUserResolveNames(Collection[str]):
     """
 
 
+class RequirementVersion(Protocol):
+    """Protocol for backend specific implementations, to support language ecosystem specific version
+    formats and sort rules.
+
+    May support the `int` properties `major`, `minor` and `micro` to color diff based on semantic
+    step taken.
+    """
+
+    def __eq__(self, other) -> bool:
+        ...
+
+    def __gt__(self, other) -> bool:
+        ...
+
+    def __lt__(self, other) -> bool:
+        ...
+
+    def __str__(self) -> str:
+        ...
+
+
+RequirementName = str
+LockfileRequirements = FrozenDict[RequirementName, RequirementVersion]
+ChangedRequirements = FrozenDict[RequirementName, Tuple[RequirementVersion, RequirementVersion]]
+
+
+@dataclass(frozen=True)
+class LockfileGenerateDiff:
```

Comment:
Do we need this class at all? There is no generic aspect of the diff generation, so this seems unnecessary.

Reply:
Strictly, no we don't. I removed it at first, but then the way it was used in the Python backend I needed it there, and figured that rather than each backend having to declare their own, we might just as well have one common class we can use across all backends (in the future, currently only the one that use it)

Reply:
Maybe I'm misunderstanding, but the Python backend does declare its own subclass of this, and I'm not sure why that class doesn't just have this field directly.

Reply:
Mm.. yea. I've clearly got lost during my refactoring. You're right, I'm more inclined to turn the python diff calculating rule into a rule helper to avoid some of all these semi-public rule types that's only used internally there. This would also help get rid of the `replace(..)` call which I find to be a code smell in this case.